File: Arrays_&_Hashing/Two_Sum/2_Better.java
import java.util.*;
class Solution {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>(nums.length);
        // First pass: fill map
        for (int i = 0; i < nums.length; i++) {
            map.put(nums[i], i);
        }
        // Second pass: search for complement
        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];
            if (map.containsKey(complement) && map.get(complement) != i) {
                return new int[]{i, map.get(complement)};
            }
        }
        return new int[]{}; // No solution
    }
}

File: Arrays_&_Hashing/Two_Sum/2_Better.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> idx;
        idx.reserve(nums.size());
        // First pass: store value -> index
        for (int i = 0; i < (int)nums.size(); ++i) {
            idx[nums[i]] = i;
        }
        // Second pass: look for complement
        for (int i = 0; i < (int)nums.size(); ++i) {
            int complement = target - nums[i];
            auto it = idx.find(complement);
            if (it != idx.end() && it->second != i) {
                return {i, it->second};
            }
        }
        return {};
    }
};
